### PR TITLE
rewrite pre-commit, post-commit and options hooks (fixes #1250)

### LINKED ISSUE #1250: hooks broken in 1.1.0
- Gitea version (or commit ref): 1.1.0
- Git version: 2.12.0
- Operating system: Linux/CentOS 7
- Database (use `[x]`):
  - [x] PostgreSQL
  - [x] MySQL
  - [x] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description

The way gitea runs hooks is broken in multiple ways.
First, as with gitea 1.1.0, the files in `pre-receive.d`, `post-receive.d` and `update.d` **must** be `sh` scripts (not even `bash`) as they are invoked using `sh $filename`. You cannot put binaries or even python scripts into these folders as `sh` cannot interpret them.
Next, no data is passed to the scripts in `post-receive.d`. So the scripts won't know about branches, revisions, you name it.

Possible fix - works as `pre-receive`, `post-receive` and `update`:
```bash
#!/usr/bin/env bash

data=$(cat)
exitcodes=()
hookname=$(basename $0)
GIT_DIR=${GIT_DIR:-$(dirname $0)}

for hook in ${GIT_DIR}/hooks/${hookname}.d/*; do
  test -x "${hook}" || continue
  echo "${data}" | "${hook}"
  exitcodes+=($?)
done

for i in "${exitcodes[@]}"; do
  [ "${i}" == 0 ] || exit ${i}
done
```
Please note that the scripts in `{{pre,post}-receive,update}.d` must have the executable bit set in order to run. So when adding custom hooks using gitea's webinterface gitea is required to set that bit.
Also note that the script above requires `bash`. If you want to have it shell-agnostic (well, at least sh-agnostic) we'll probably have to rewrite the `exitcodes`-array-stuff.

### PR DESCRIPTION
**Pull request by philfry**
see #1250

### CODE REVIEW

**Review by strk — CHANGES_REQUESTED** [review]
I see you're feeding a bash script to a user-configurable `scriptType`, sounds fragile to me. Can you us a POSIX syntax please ?

File: models/repo.go [hunk 1]
```diff
@@ -845,11 +845,7 @@ func cleanUpMigrateGitConfig(configPath string) error {
...
 	var (
 		hookNames = []string{"pre-receive", "update", "post-receive"}
-		hookTpls  = []string{
-			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/pre-receive.d\"`; do\n    sh \"$SHELL_FOLDER/pre-receive.d/$i\"\ndone", setting.ScriptType),
-			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/update.d\"`; do\n    sh \"$SHELL_FOLDER/update.d/$i\" $1 $2 $3\ndone", setting.ScriptType),
-			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/post-receive.d\"`; do\n    sh \"$SHELL_FOLDER/post-receive.d/$i\"\ndone", setting.ScriptType),
-		}
+		hookTpl = fmt.Sprintf("#!/usr/bin/env %s\ndata=$(cat)\nexitcodes=()\nhookname=$(basename $0)\nGIT_DIR=${GIT_DIR:-$(dirname $0)}\n\nfor hook in ${GIT_DIR}/hooks/${hookname}.d/*; do\ntest -x \"${hook}\" || continue\necho \"${data}\" | \"${hook}\"\nexitcodes+=($?)\ndone\n\nfor i in \"${exitcodes[@]}\"; do\n[ \"${i}\" == 0 ] || exit ${i}\ndone\n", setting.ScriptType)
```
Comment (strk) [inline models/repo.go]: bash-specific syntax (`$(cat)`) in what could be running with something else (`setting.ScriptType`).
Not that I understand how can Gitea possibly contains code to be run by unknown scripts, anyway the app.ini Cheat Sheet says `ScriptType` is usually bash or sh...

**Review by strk — CHANGES_REQUESTED** [review]

**Comment by vtolstov** [comment]
Why we can't create symlink for hooks to gitea binary? In this case we
don't need bash, sh and other stuff. And gitea handle hook internally.

15 Мар 2017 г. 9:29 пользователь "Sandro Santilli" <notifications@github.com>
написал:

File: models/repo.go (hunk 1 above)
Comment (philfry) [reply models/repo.go]: `data=$(cat)`  is not bash-specific.

`bash` can be run in posix mode when either invoking the script either using `sh` (`sh scriptname`, `sh` in shebang) or by using the switch `--posix` (or using `set -o posix` in the script itself). Tried both of it, script worked fine. Even though `myarray=()` is a bashism.

Some Debian systems might symlink `/bin/sh` to `/bin/dash` (the Debian Almquist Shell), so we might want to force using `bash` in this case.

Or, to make it more posix-compliant, give up on arrays:
```sh
#!/bin/sh

data=$(cat)
exitcodes=""
hookname=$(basename $0)
GIT_DIR=${GIT_DIR:-$(dirname $0)}

for hook in ${GIT_DIR}/hooks/${hookname}.d/*; do
	test -x "${hook}" || continue
	echo "${data}" | "${hook}"
	exitcodes="${exitcodes} $?"
done

for i in ${exitcodes}; do
	[ $i -eq 0 ] || exit $i
done
```
(tested with `dash`)

**Comment by strk** [comment]
I dunno why are you talking about symlinks, but yes I think we could just invoke the commands, if executables. But please make sure you touch all files containing that bash invocation, I see 3 places.
Also it would be great to have unit tests for this

**Comment by philfry** [comment]
Am I missing something? The `migrations`?
```
$ find . -name '*.go' -print0 | xargs -r0 grep -P '^(?!//).*\bsh\b'
./models/migrations/v22.go:			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/pre-receive.d\"`; do\n    sh \"$SHELL_FOLDER/pre-receive.d/$i\"\ndone", setting.ScriptType),
./models/migrations/v22.go:			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/update.d\"`; do\n    sh \"$SHELL_FOLDER/update.d/$i\" $1 $2 $3\ndone", setting.ScriptType),
./models/migrations/v22.go:			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/post-receive.d\"`; do\n    sh \"$SHELL_FOLDER/post-receive.d/$i\"\ndone", setting.ScriptType),
./models/migrations/v19.go:			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/pre-receive.d\"`; do\n    sh \"$SHELL_FOLDER/pre-receive.d/$i\"\ndone", setting.ScriptType),
./models/migrations/v19.go:			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/update.d\"`; do\n    sh \"$SHELL_FOLDER/update.d/$i\" $1 $2 $3\ndone", setting.ScriptType),
./models/migrations/v19.go:			fmt.Sprintf("#!/usr/bin/env %s\nORI_DIR=`pwd`\nSHELL_FOLDER=$(cd \"$(dirname \"$0\")\";pwd)\ncd \"$ORI_DIR\"\nfor i in `ls \"$SHELL_FOLDER/post-receive.d\"`; do\n    sh \"$SHELL_FOLDER/post-receive.d/$i\"\ndone", setting.ScriptType),
./vendor/golang.org/x/text/language/maketables.go:	"ro", "ru", "sh", "si", "sk", "sl", "sq", "sr", "sv", "sw", "ta", "te", "th",
./vendor/golang.org/x/text/language/maketables.go:	lang.updateLater("sh", "hbs")
./vendor/golang.org/x/sys/unix/types_openbsd.go:Input to cgo -godefs.  See also mkerrors.sh and mkall.sh
./vendor/golang.org/x/sys/unix/types_darwin.go:Input to cgo -godefs.  See also mkerrors.sh and mkall.sh
./vendor/golang.org/x/sys/unix/types_netbsd.go:Input to cgo -godefs.  See also mkerrors.sh and mkall.sh
./vendor/golang.org/x/sys/unix/types_freebsd.go:Input to cgo -godefs.  See also mkerrors.sh and mkall.sh
./vendor/golang.org/x/sys/unix/types_linux.go:Input to cgo -godefs.  See also mkerrors.sh and mkall.sh
./vendor/golang.org/x/sys/unix/types_solaris.go:Input to cgo -godefs.  See also mkerrors.sh and mkall.sh
./vendor/golang.org/x/sys/unix/types_dragonfly.go:Input to cgo -godefs.  See also mkerrors.sh and mkall.sh
./modules/base/tool.go:	sh := sha1.New()
./modules/base/tool.go:	sh.Write([]byte(data + setting.SecretKey + startStr + endStr + com.ToStr(minutes)))
./modules/base/tool.go:	encoded := hex.EncodeToString(sh.Sum(nil))
./modules/highlight/highlight.go:		".sh":     true,
```

**Comment by strk** [comment]
Yes, the migrations. I'm not very sure what are them for, and how to effectively test them, will be tricky.
You can always trigger them by setting version.version in the database to the lowest version (19, in this case) and starting Gitea again. I guess it takes installing some hooks in the old format to test the migrations.

**Comment by lunny** [comment]
The migrations will be simple to rewrite all the scripts. But don't forget wiki repo's hooks.

**Comment by philfry** [comment]
afaics the wiki pages use the same hooks as the repos do. Which leads to funny side-effects[1] but that's not relevant here.
So, nothing to do for the wikis here.

I'm a little concerned about the migration scripts. In my understanding the scripts should update database schemas, scripts (i.e. hooks) and other stuff that might have changed between gitea versions.
I can change `v19.go` and `v22.go` to match the new hook scripts. So when updating from gitea .. say 0.9 .. the hooks will be rewritten. But when updating from gitea 1.1.0  to gitea > 1.1.0 will the changes be applied? Or do we need a completely new migration script `v23.go` (or whatever it will be named) that replaces the hooks once again?


**edit**
I think now I understand how these migration scripts work. Every migration script gets called on every update, regardless of its "version" number. So I'll only have to change some conditionals in `v19.go` and `v22.go` like `if com.IsExist(customHooksDir)`.



[1] I have `git-multimail.py` in my git templates so on every wiki page change I get a mail with the diff :)

**Comment by strk** [comment]
Good question about migrations. I think first step is understanding what those migrations did in `v19` and `v22`. If they installed broken scripts, now we'd need a new migration to fix those scripts (if we can ensure those scripts are "owned" by gitea)

**Comment by philfry** [comment]
the migration works for me.
Tested:
- freshly installed gitea 1.0.0 → 1.1.0 → 1.1.0fix_hooks
- freshly installed gitea 1.1.0 → 1.1.0fix_hooks
- gitea 1.1.0 w/ custom hooks → 1.1.0fix_hooks

**Comment by bkcsoft** [comment]
As @strk pointed out, don't touch existing migrations, it will break things even worse. Create a new migration that updates all existing hooks.

**Comment by philfry** [comment]
@bkcsoft I did not touch existing migrations.

File: models/migrations/v26.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
+package migrations
```
Comment (lunny) [inline models/migrations/v26.go:5]: comment header

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/migrations/migrations.go [hunk 3]
```diff
@@ -100,6 +100,8 @@ var migrations = []Migration{
 	NewMigration("change the key_id and primary_key_id type", changeGPGKeysColumns),
 	// v25 -> v26
 	NewMigration("add show field in user openid table", addUserOpenIDShow),
+  // v26 -> v27
+  NewMigration("generate and migrate repo and wiki Git hooks", generateAndMigrateGitHookChains),
```
Comment (lunny) [inline models/migrations/migrations.go]: Format

**Comment by lunny** [comment]
And maybe you can also fix another issue
```
SyncRepositoryHooks
```
This function didn't sync wiki repository's hooks.

**Comment by lunny** [comment]
otherwise LGTM

File: models/migrations/migrations.go (hunk 3 above)
Comment (strk) [reply models/migrations/migrations.go]: Run `make fmt` should be enough (and we need our bot to guard after this, I may give that a try)

**Comment by strk** [comment]
So doesn't it take any migration to fix the existing hooks ? How/when do those wrapper scripts get installed in git repos ?

**Comment by philfry** [comment]
@strk I'm not sure what you're talking about. Actually there _is_ a migration script (`v26.go`) to install the new hooks.

@lunny ~I'll try to fix it. Do you want it in this PR or shall I open a new PR when done?~ (edit: the change is tiny enough to have it in this PR)
btw - the "comment header" change is done but this PRs status is still on "Changes requested".

**Review by lunny — APPROVED** [review]

**Comment by strk** [comment]
@philfry sorry I think I was fooled by github as when I wrote that comment I really didn't see the migration file !

LGTM (trusted, didn't really test it)

**Comment by lunny** [comment]
@philfry, could you send a backport PR to v1.1? But maybe the migrations is difficult to handle.

**Pull request merged**